### PR TITLE
V2 controller metrics concurrency fix

### PR DESCRIPTION
**Pull request by shaoxt**
Fix the v2 controller metrics concurrency issue

### CODE REVIEW

**Review by zhatt — COMMENTED** [review]
Is there some way the mutex can be in the RollingUpgradeStatus struct instead of passing it around as part of the API?

**Comment by eytan-avisror** [comment]
+1 for embedding the mutex in a struct instead of passing it.
I think this entire implementation should not be part of `RollingUpgradeStatus` TBH
metrics should be part of reconciler and configure status from within the context of a reconcile

**Comment by shaoxt** [comment]
> Is there some way the mutex can be in the RollingUpgradeStatus struct instead of passing it around as part of the API?

That was the first approach I did.  But the generator does not allow it.

**Comment by shaoxt** [comment]
> +1 for embedding the mutex in a struct instead of passing it.
> I think this entire implementation should not be part of `RollingUpgradeStatus` TBH
> metrics should be part of reconciler and configure status from within the context of a reconcile
The metrics are not only shown in metrics but also on the CR.
That's the reason it has part of RollingUpgradeStatus.

Regarding the mutex.  When I put the following struct in CR. Controller-gen doesn't allow it.

```
type NodeStepMap struct {
	mutex *sync.Mutex

	nodeSteps map[string][]NodeStepDuration
}
```

```
# github.com/keikoproj/upgrade-manager/api/v1alpha1
api/v1alpha1/zz_generated.deepcopy.go:90:11: assignment copies lock value to **out: sync.Mutex
make: *** [vet] Error 2
```

**Comment by sahilbadla** [comment]
@shaoxt is [Sync Map ](https://golang.org/pkg/sync/)a better option here then?

**Comment by eytan-avisror** [comment]
@shaoxt this is because you are putting the implementation inside the API (status) and it gets copied by `zz_generated.deepcopy` - this is why I suggested to move the entire metrics implementation into the context of a RollingUpgrade, and not part of the CR API Objects itself.

Right now we only have a Controller level object `RollingUpgradeReconciler` which holds state that is common to all CRs - this is why you could not find a good place to put it. @shreyas-badiger has a refactor that adds another level `RollingUpgradeContext` which holds state for the individual resource - this would be an idea place to store metrics.

Can you wait for him to merge that PR first? I asked him to send it. It would then be much easier to move this code and embed a mutex.

**Comment by eytan-avisror** [comment]
> @shaoxt is [Sync Map ](https://golang.org/pkg/sync/)a better option here then?

Sync map would work (considering you can copy it - otherwise it would get the same error from `zz_generated.deepcopy`), but it would mean we have to change wherever we access it, add casting, etc + iteration becomes a pain vs. simple as a map with mutex.

**Comment by shaoxt** [comment]
> > @shaoxt is [Sync Map ](https://golang.org/pkg/sync/)a better option here then?
> 
> Sync map would work (considering you can copy it - otherwise it would get the same error from `zz_generated.deepcopy`), but it would mean we have to change wherever we access it, add casting, etc + iteration becomes a pain vs. simple as a map with mutex.

I think moving the method into RollingUpgradeContext makes sense.   Yes.  I considered the SyncMap at the beginning. But have to change all the argument type.

**Comment by eytan-avisror** [comment]
It's merged now @shaoxt if you want to move the implementation

**Comment by shaoxt** [comment]
@eytan-avisror Refined the code based on RollingUpgradeContext

File: controllers/upgrade.go [hunk 1]
```diff
@@ -109,6 +111,110 @@ func (r *RollingUpgradeContext) RotateNodes() error {
 	return nil
 }
 
+// Update last batch nodes
+func (s *RollingUpgradeContext) UpdateLastBatchNodes(batchNodes map[string]*v1alpha1.NodeInProcessing) {
```
Comment (eytan-avisror) [inline controllers/upgrade.go:115]: Can we move all these methods to `controllers/metrics.go`?

File: controllers/upgrade_test.go [hunk 2]
```diff
@@ -385,3 +386,48 @@ func TestDesiredNodesReady(t *testing.T) {
 		}
 	}
 }
+
+// Test
+func TestNodeTurnsOntoStep(t *testing.T) {
```
Comment (eytan-avisror) [inline controllers/upgrade_test.go:391]: Also this to `controllers/metrics_test.go`

File: controllers/upgrade.go (hunk 1 above)
Comment (shaoxt) [reply controllers/upgrade.go:115]: package "api" depends on package "common". 
We can not let package "common" refer to the structures in "api.*".
Also common is used in "controllers", those methods need to access RollingUpgradeContext.

Comment (eytan-avisror) [reply controllers/upgrade.go:115]: This can stay same package @shaoxt 
just different file

Comment (shaoxt) [reply controllers/upgrade.go]: Sounds good to me.  Moved. Thanks

File: controllers/upgrade_test.go (hunk 2 above)
Comment (shaoxt) [reply controllers/upgrade_test.go]: Moved to upgrade_metrics_test.go

**Review by eytan-avisror — APPROVED** [review]

**Review by shreyas-badiger — APPROVED** [review]

**Pull request merged**